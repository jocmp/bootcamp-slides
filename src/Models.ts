export interface SlideshowModel {
    id: number;
    title: string;
    slides: SlideModel[];
}

export interface SlideModel {
    title: string;
    slide_type: string;
    content?: string[];
}

export interface AppState {
    slideshow: SlideshowModel;
}