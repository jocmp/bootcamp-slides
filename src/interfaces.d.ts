declare module 'enzyme-to-json'