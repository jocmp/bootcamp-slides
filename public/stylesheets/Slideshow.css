.slideshow-container {
  width: 55%;
  margin-right: 20%;
  display: block;
  float: left; }

.slideshow {
  height: 60vh;
  vertical-align: center;
  margin-top: 10px;
  border: 1px solid #4C4845; }

.directional-button, .next-button, .previous-button {
  display: inline-block;
  padding: 12px 20px;
  font-size: 12pt;
  cursor: pointer;
  text-align: center;
  text-decoration: none;
  outline: none;
  color: #fff;
  background-color: #A5488B;
  border: none;
  border-radius: 12px;
  box-shadow: 0 8px #999;
  margin-top: 8px; }
  .directional-button:active, .next-button:active, .previous-button:active {
    background-color: #7F386B;
    box-shadow: 0 5px #666;
    transform: translateY(4px); }
  .directional-button:hover, .next-button:hover, .previous-button:hover {
    background-color: #7F386B; }

.next-button {
  background-color: #16CBC4;
  float: right; }
  .next-button:hover {
    background-color: #0E7F7B; }

.previous-button {
  float: left;
  background-color: white;
  border: 1px solid #CCC;
  color: #333; }
  .previous-button:hover {
    background-color: #E6E6E6; }
