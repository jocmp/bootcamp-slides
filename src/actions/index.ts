import { SlideshowModel } from '../Models'
import { History } from 'history';

export const nextSlide = (id: string, index: string, totalLength: number, history: History) => {
    const currentIndex = parseInt(index);
    const nextIndex: number = currentIndex < totalLength - 1 ? currentIndex + 1 : currentIndex;
    return handleSlide(id, nextIndex, history);
}

export const previousSlide = (id: string, index: string, totalLength: number, history: History) => {
    const currentIndex = parseInt(index);
    const previousIndex: number = currentIndex > 0 ? currentIndex - 1 : currentIndex;
    return handleSlide(id, previousIndex, history);
}

const handleSlide = (id: string, index: number, history: History) => {
    history.push(`/slideshows/${id}/slides/${index}`);
    return viewSlide(index);
}

export const viewSlide = (index: number) => {
    return {
        type: 'VIEW_SLIDE',
        index
    };
}

const apiData = (index: number): SlideshowModel => {
    const shows: SlideshowModel[] = [
        {
            id: 1,
            title: "Exceptional Presentation",
            slides: [
                {
                    title: "Title here",
                    slide_type: "title"
                },
                {
                    title: "Sensational Informational",
                    slide_type: "simple",
                    content: ["Inspirational insatiable"]
                }
            ]
        },
        {
            id: 2,
            title: "Touch Fuzzy, Get Dizzy",
            slides: [
                {
                    title: "Yoshi's Island",
                    slide_type: "title",
                },
                {
                    title: "Yoshi's Island",
                    slide_type: "simple",
                    content: ['Imperically the best Super Mario game... \ntest']
                },
                {
                    title: "Why buy a SNES?",
                    slide_type: "simple",
                    content: ["Yoshi's Island. No question."]
                }
            ]
        }
    ]
    return shows[index - 1];
};

export const loadSlideshow = (slideshow: SlideshowModel, currentIndex: number) => {
    return {
        type: 'LOAD_SLIDESHOW',
        slideshow,
        currentIndex
    }
};

export const fetchSlideshow = (id: number, index: number = 0) => (dispatch: any, getState: any) => {
    dispatch(loadSlideshow(apiData(id), index));
};