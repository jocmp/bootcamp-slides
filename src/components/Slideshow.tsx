import * as React from 'react'
import Slide from './Slide'

const slideIndices: number[] = [1, 2, 3]
const Slideshow = () => {
    return (
        <ul>
        {slideIndices.map(index =>
            <Slide slideIndex={index}/>
        )}
        </ul>
    );
}

export default Slideshow;