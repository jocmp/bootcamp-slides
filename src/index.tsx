import * as React from 'react';
import * as ReactDOM from 'react-dom';
import App from './components/App';
import './index.css';
import { createStore, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'

export const slideshow = (state: any, action: any) => {
  switch(action.type) {
    case 'LOAD_SLIDESHOW':
      return {
        slideshow: action.slideshow
      }
    default:
      return {};
  }
}

export const store = createStore(slideshow, applyMiddleware(thunk));
store.subscribe(() => console.log(store.getState()))

ReactDOM.render(
  <App store={store} />,
  document.getElementById('root') as HTMLElement
);
