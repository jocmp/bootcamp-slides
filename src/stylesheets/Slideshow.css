.slideshow-container {
  width: 50%;
  margin-left: 25%;
  margin-right: 25%; }

.slideshow {
  height: 60vh;
  vertical-align: center;
  margin-top: 10px;
  border: 1px solid black; }

.next-button {
  float: right; }

.previous-button {
  float: left; }
