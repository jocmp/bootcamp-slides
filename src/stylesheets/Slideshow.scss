@import "./colors";

$button-background: #A5488B;
$button-background-highlighted: #7F386B;

.slideshow-container {
  width: 55%;
  // background: pink;
  margin-right: 20%;
  display: block;
  float: left;
}

.slideshow {
  height: 60vh;
  vertical-align: center;
  margin-top: 10px;
  border: 1px solid $black;
 }

.directional-button {
  // https://www.w3schools.com/css/tryit.asp?filename=trycss_buttons_animate3
  display: inline-block;
  padding: 12px 20px;
  font-size: 12pt;
  cursor: pointer;
  text-align: center;
  text-decoration: none;
  outline: none;
  color: #fff;
  background-color: $button-background; 
  border: none;
  border-radius: 12px;
  box-shadow: 0 8px #999;
  margin-top: 8px;

  &:active {
    background-color: $button-background-highlighted;
    box-shadow: 0 5px #666;
    transform: translateY(4px);
  }

  &:hover {
    background-color: $button-background-highlighted;
  }
}

.next-button {
  @extend .directional-button;
  background-color: $teal;
  float: right;

  &:hover {
    background-color: $teal-highlighted; 
  }
}

.previous-button {
  @extend .directional-button;
  float: left;
  background-color: white;
  border: 1px solid #CCC;
  color: #333;

  &:hover {
    background-color: #E6E6E6;
  }
}

.invisible {
  visibility: hidden;
}