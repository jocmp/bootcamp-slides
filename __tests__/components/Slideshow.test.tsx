import * as React from 'react'
import Slideshow from '../../src/components/Slideshow'
import { SlideshowModel, SlideModel } from '../../src/Models'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

describe('Slideshow', () => {

    const slide: SlideModel = {
        title: "Exciting Slide",
        slide_type: "title",
    };
    const show: SlideshowModel = {
        id: 1,
        title: "Exciting Presentation",
        slides: [slide, slide, slide]
    };

    const viewed: number[] = [];

    it('renders', () => {
        const wrapper = shallow(
            <Slideshow slideshow={show} viewedIndices={viewed}
                match={{ params: {} }} handleNextSlide={jest.fn()} viewSlide={jest.fn()} handlePreviousSlide={jest.fn()} history={history} fetchSlideshow={(id) => show} />)
        expect(toJson(wrapper)).toMatchSnapshot()
    });

    describe('next slide event', () => {
        it('responds to next click', () => {
            const nextListener = jest.fn();

            const wrapper = shallow(<Slideshow slideshow={show} match={{ params: { index: 1 } }} viewedIndices={viewed}
                handleNextSlide={nextListener} viewSlide={jest.fn()} handlePreviousSlide={jest.fn()} history={history} fetchSlideshow={(id: number) => show} />)
            wrapper.find('.next-button').simulate('click') // optionally: stubbed out text edit event
            expect(nextListener).toHaveBeenCalled();
        });
    });

    describe('previous slide event', () => {
        it('responds to previous click', () => {
            const previousListener = jest.fn();

            const wrapper = shallow(<Slideshow slideshow={show} match={{ params: { index: 2 } }} viewedIndices={viewed}
                handleNextSlide={jest.fn()} viewSlide={jest.fn()} handlePreviousSlide={previousListener} history={history} fetchSlideshow={(id: number) => show} />)
            wrapper.find('.previous-button').simulate('click');
            expect(previousListener).toHaveBeenCalled();
        });
    });
});