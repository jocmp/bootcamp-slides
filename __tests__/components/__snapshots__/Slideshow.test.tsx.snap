// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slideshow renders 1`] = `
<div>
  <Overview
    slideshow={
      Object {
        "id": 1,
        "slides": Array [
          Object {
            "slide_type": "title",
            "title": "Exciting Slide",
          },
          Object {
            "slide_type": "title",
            "title": "Exciting Slide",
          },
          Object {
            "slide_type": "title",
            "title": "Exciting Slide",
          },
        ],
        "title": "Exciting Presentation",
      }
    }
    viewedIndices={Array []}
  />
  <div
    className="slideshow-container"
  >
    <div
      className="slideshow"
    />
    <button
      className="invisible"
    >
      Previous
    </button>
    <button
      className="invisible"
    >
      Next
    </button>
  </div>
</div>
`;
